### PR TITLE
Multi-head CRNN

### PR DESCRIPTION
**Pull request by themattinthehatt**

### CODE REVIEW

File: lightning_pose/models/base.py [hunk 1]
```diff
@@ -326,16 +314,27 @@ def configure_optimizers(self) -> dict:
...
-            )
+            raise NotImplementedError("'%s' is an invalid LR scheduler" % self.lr_scheduler)
+
+        return scheduler
+
+    def configure_optimizers(self) -> dict:
+        """Select optimizer, lr scheduler, and metric for monitoring."""
+
+        # get trainable params
+        params = self.get_parameters()
```
Comment (danbider) [inline lightning_pose/models/base.py:331]: I hope that this command works fine with the order of the parameters and the learning rate decays.

File: lightning_pose/models/heatmap_tracker.py [hunk 2]
```diff
@@ -202,10 +201,26 @@ def create_double_upsampling_layer(
...
+        """Handle context frames then upsample to get final heatmaps."""
+
+        # handle context frames first
+        if (self.mode == "2d" and self.do_context) or self.mode == "3d":
+            # push through a linear layer to get the final representation
+            # input shape (batch, features, rep_height, rep_width, frames)
+            # output shape (batch, features, rep_height, rep_width, 1)
+            representations = self.representation_fc(representations)
+            # final squeeze
+            # output shape (batch, features, rep_height, rep_width)
```
Comment (danbider) [inline lightning_pose/models/heatmap_tracker.py:218]: maybe instead of comment, can have 
```
representations: TensorType["batch", "features", "rep_height", "rep_width"] = torch.squeeze(representations, 4)
```

File: lightning_pose/models/heatmap_tracker_mhcrnn.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+            params.append({
+                "params": self.loss_factory_unsup.loss_weights_parameter_dict.parameters(),
+                "lr": 1e-2,
+            })
+        return params
+
+
+class UpsamplingCRNN(torch.nn.Module):
+
+    def __init__(
```
Comment (danbider) [inline lightning_pose/models/heatmap_tracker_mhcrnn.py:303]: more typechecking?
might be good if we add a formula in a comment that links those W's and H's.

File: lightning_pose/models/heatmap_tracker_mhcrnn.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+            self, num_filters_for_upsampling, num_keypoints, upsampling_factor=2, hkernel=2,
+            hstride=2, hpad=0, nfilters_channel=16):
+        """Upsampling Convolutional RNN - initialize input and hidden weights."""
+
+        super().__init__()
+        self.upsampling_factor = upsampling_factor
+        self.pixel_shuffle = nn.PixelShuffle(2)
+        if self.upsampling_factor == 2:
```
Comment (danbider) [inline lightning_pose/models/heatmap_tracker_mhcrnn.py:311]: maybe can comment here:
`if self.upsampling_factor == 2: # need to upsample representations twice`

`else: # representations will be upsampled once, in following lines`

File: lightning_pose/models/regression_tracker.py [hunk 5]
```diff
@@ -113,9 +102,17 @@ def forward(
...
         representations = self.get_representations(images)
+        # handle context frames first
+        if (self.mode == "2d" and self.do_context) or self.mode == "3d":
+            # push through a linear layer to get the final representation
+            # input shape (batch, features, rep_height, rep_width, frames)
+            # output shape (batch, features, rep_height, rep_width, 1)
+            representations = self.representation_fc(representations)
+            # final squeeze
+            # output shape (batch, features, rep_height, rep_width)
+            representations = torch.squeeze(representations, 4)
```
Comment (danbider) [inline lightning_pose/models/regression_tracker.py:113]: same comment as above -- can do typedtensor

File: lightning_pose/utils/predictions.py [hunk 6]
```diff
@@ -515,35 +515,29 @@ def get_model_class(map_type: str, semi_supervised: bool) -> Type[LightningModul
...
                 "%s is an invalid model_type for a fully supervised model" % map_type
             )
     else:
         if map_type == "regression":
-            from lightning_pose.models.regression_tracker import (
-                SemiSupervisedRegressionTracker,
-            )
-
-            return SemiSupervisedRegressionTracker
+            from lightning_pose.models.regression_tracker import SemiSupervisedRegressionTracker as Model
```
Comment (danbider) [inline lightning_pose/utils/predictions.py:529]: nice import

**Review by danbider — CHANGES_REQUESTED** [review]
minor changes only, we can merge very quickly

File: lightning_pose/models/base.py (hunk 1 above)
Comment (themattinthehatt) [reply lightning_pose/models/base.py:331]: I know that it does because I had a bug in this part of the code for a while - I accidentally named the inherited function `get_params` in the mhcrnn at first, and was getting terrible training results (because the crnn wasn't added to list of trainable params). once i fixed that model training worked beautifully.

File: lightning_pose/models/heatmap_tracker_mhcrnn.py (hunk 3 above)
Comment (themattinthehatt) [reply lightning_pose/models/heatmap_tracker_mhcrnn.py:303]: added typechecking, we can ask cole to add a formula later

File: lightning_pose/models/heatmap_tracker_mhcrnn.py (hunk 4 above)
Comment (themattinthehatt) [reply lightning_pose/models/heatmap_tracker_mhcrnn.py:311]: this is the first time i actually looked at this part of the code, i rewrote it to be easier to parse

Comment (themattinthehatt) [reply lightning_pose/models/heatmap_tracker_mhcrnn.py:311]: (and tested that it gave the same results)

**Review by danbider — APPROVED** [review]
